File: frontend/src/types/invoice.ts
/** Mirrors backend/app/schemas/invoice.py. */

export type InvoiceStatus =
  | 'draft'
  | 'waiting_finance'
  | 'link_generated'
  | 'qr_ready'
  | 'awaiting_payment'
  | 'partially_paid'
  | 'paid'
  | 'voided'

/**
 * What everyone with "orders" read access sees -- a scoped salesman, the
 * Sales Manager, or anyone else who can see the underlying order. Never
 * the payment link, QR, or MyFatoorah reference -- those are Finance's
 * (see InvoiceFull below), not Sales' to even read.
 */
export interface Invoice {
  id: number
  invoice_number: string
  order_id: number
  order_number: string | null
  quotation_id: number | null
  quotation_number: string | null
  customer_id: number
  customer_name: string | null
  status: InvoiceStatus
  total_amount: number
  amount_acknowledged: number
  voided_reason: string | null
  created_at: string
  updated_at: string
}

/** Finance's full view (permissions.payments === 'write') -- adds the
 * payment link, QR, and MyFatoorah reference InvoiceStatusOut withholds. */
export interface InvoiceFull extends Invoice {
  version: number
  payment_link_url: string | null
  payment_link_ref: string | null
  payment_link_expires_at: string | null
  qr_data_url: string | null
  voided_at: string | null
  voided_by: number | null
}

export const INVOICE_STATUS_LABELS: Record<InvoiceStatus, string> = {
  draft: 'Draft',
  waiting_finance: 'Waiting for Finance',
  link_generated: 'Link generated',
  qr_ready: 'QR ready',
  awaiting_payment: 'Awaiting payment',
  partially_paid: 'Partially paid',
  paid: 'Payment received',
  voided: 'Voided',
}

File: backend/app/schemas/invoice.py
from datetime import datetime

from pydantic import BaseModel, Field


class InvoiceVoidIn(BaseModel):
    reason: str = Field(min_length=1, max_length=2000)


class InvoiceStatusOut(BaseModel):
    """What Sales (a scoped salesman or the Sales Manager) may see: the
    invoice exists and where it stands, never the payment link, QR, or
    MyFatoorah reference -- those are Finance's to edit, not Sales' to even
    read (see the "Sales -> Finance Invoice Handoff" design doc, gap 7).
    """

    id: int
    invoice_number: str
    order_id: int
    order_number: str | None = None
    quotation_id: int | None
    quotation_number: str | None = None
    customer_id: int
    customer_name: str | None = None
    status: str
    # Not columns on Invoice itself (order total / SUM of acknowledged
    # payments) -- left at their default here so model_validate(obj) can
    # build the rest of the object straight off the ORM row, then
    # from_model below fills these in from what the caller already
    # computed (see OrderOut.from_model's own precomputed-extras pattern
    # -- schemas don't reach into services themselves).
    total_amount: float = 0.0
    amount_acknowledged: float = 0.0
    voided_reason: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @staticmethod
    def from_model(obj, amount_acknowledged: float = 0.0) -> "InvoiceStatusOut":
        data = InvoiceStatusOut.model_validate(obj)
        data.order_number = obj.order.order_number if obj.order else None
        data.quotation_number = obj.quotation.quotation_number if obj.quotation else None
        data.customer_name = obj.customer.name if obj.customer else None
        data.total_amount = float(obj.order.total_amount) if obj.order else 0.0
        data.amount_acknowledged = amount_acknowledged
        return data


class InvoiceOut(InvoiceStatusOut):
    """Finance's full view -- adds the payment link, QR, and MyFatoorah
    reference that InvoiceStatusOut deliberately withholds."""

    version: int
    payment_link_url: str | None
    payment_link_ref: str | None
    payment_link_expires_at: datetime | None
    qr_data_url: str | None
    voided_at: datetime | None
    voided_by: int | None

    @staticmethod
    def from_model(obj, amount_acknowledged: float = 0.0) -> "InvoiceOut":
        data = InvoiceOut.model_validate(obj)
        data.order_number = obj.order.order_number if obj.order else None
        data.quotation_number = obj.quotation.quotation_number if obj.quotation else None
        data.customer_name = obj.customer.name if obj.customer else None
        data.total_amount = float(obj.order.total_amount) if obj.order else 0.0
        data.amount_acknowledged = amount_acknowledged
        return data
